File: leetcode/2-sum/2-sum.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {

	//map with numbers as keys and their indexes as values
	numbers := make(map[int][]int)
	//var numbers map[int][]int

	// generate the map
	for idx, num := range nums {
		if _, ok := numbers[num]; !ok {
			numbers[num] = []int{idx} // add first position of num to map

			fmt.Println(numbers[num])
		} else {
			numbers[num] = append(numbers[num], idx) // add the other indexes of the number
		}
	}

	for _, num := range nums {
		if _, ok := numbers[target-num]; ok {

			//if numbers are equal, we must have exactly two
			if target-num == num {
				if len(numbers[num]) == 1 {
					continue
				} else {
					return numbers[num]
				}
			} else {
				return []int{numbers[num][0], numbers[target-num][0]}
			}
		}
	}
	return nil
}

//Not really worth the improvement
func twoSumOnePass(nums []int, target int) []int {
	// same solution as the previous one, do it in one pass

	//map with numbers as keys and their indexes as values
	numbers := make(map[int][]int)
	//var numbers map[int][]int

	// generate the map
	for idx, num := range nums {
		if _, ok := numbers[num]; !ok {
			numbers[num] = []int{idx} // add first position of num to map)

			if _, ok := numbers[target-num]; ok {

				//if numbers are equal, we must have exactly two
				if target-num == num {
					if len(numbers[num]) == 1 {
						continue
					} else {
						return numbers[num]
					}
				} else {
					return []int{numbers[num][0], numbers[target-num][0]}
				}
			}

		} else {
			numbers[num] = append(numbers[num], idx) // add the other indexes of the number
			if _, ok := numbers[target-num]; ok {

				//if numbers are equal, we must have exactly two
				if target-num == num {
					if len(numbers[num]) == 1 {
						continue
					} else {
						return numbers[num]
					}
				} else {
					return []int{numbers[num][0], numbers[target-num][0]}
				}
			}
		}
	}
	return nil
}

func main() {
	fmt.Println(twoSum([]int{2, 7, 11, 15}, 9))
}

File: leetcode/2-sum/2-sum.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # note that we can even do this in a single attempt, but it's probably not worth the imporvement

        numbers_dict = {} # we will use a dictionary to store the numbers as keys, and their indexes as values (list)
        for idx, n in enumerate(nums):
            if n not in numbers_dict:
                numbers_dict[n] = [idx]
            else:
                # if the number is already in the hashmap, it's present with a different index as well
                numbers_dict[n].append(idx)

        for number in nums:
            if target - number in numbers_dict: # basically check if the sum is target
                if target - number == number:
                    # if the numbers are equal, then we must have exactly two of them to ensure that we didn't add twice
                    if len(numbers_dict[number]) == 1:
                        continue
                    else:
                        return numbers_dict[number] # since the solution is unique
                else:
                    return [numbers_dict[number][0], numbers_dict[target - number][0]]


s = Solution()
print(s.twoSum(nums = [3,2,4], target = 6))